packages/core/src/dialects/abstract/query-generator-internal.ts

```diff
@@ -305,11 +315,99 @@ Only named replacements (:name) are allowed in literal() because we cannot guara
    *
    * @param _options
    */
   addLimitAndOffset(_options: AddLimitOffsetOptions): string {
     throw new Error(`addLimitAndOffset has not been implemented in ${this.dialect.name}.`);
   }
+
+  /**
+   * Returns a SQL fragment for a WHERE with a LIMIT/OFFSET using a subquery.
+   *
+   * @param tableName
+   * @param options
+   */
+  addWhereSelectLimitOffset(tableName: TableNameOrModel, options: AddWhereSelectLimitOffset): string {
+    const model = isModelStatic(tableName) ? tableName : options.model;
+    const whereOptions = { ...options, model };
+    const whereQuery = whereOptions.where ? this.queryGenerator.whereQuery(whereOptions.where, whereOptions) : '';
+    if (whereOptions.limit && this.dialect.supports.modelWithLimit) {
+      if (!model) {
+        throw new Error('Cannot use LIMIT without a model.');
+      }
+
+      const pks = Object.values(model.primaryKeys).map(key => this.queryGenerator.quoteIdentifier(key.columnName)).join(', ');
+      const primaryKeys = Object.values(model.primaryKeys).length > 1 ? `(${pks})` : pks;
+
+      return joinSQLFragments([
+        `WHERE ${primaryKeys} IN (`,
+        `SELECT ${pks} FROM ${this.queryGenerator.quoteTable(tableName)}`,
+        whereQuery,
+        `ORDER BY ${pks}`,
+        this.addLimitAndOffset(whereOptions),
+        ')',
+      ]);
+    }
+
+    return joinSQLFragments([
+      whereQuery,
+      this.addLimitAndOffset(whereOptions),
+    ]);
+  }
+
+  /**
+   * Creates a function that can be used to collect bind parameters.
+   *
+   * @param bind A mutable object to which bind parameters will be added.
+   */
+  bindParam(bind: Record<string, unknown>): (value: unknown) => string {
+    let i = 0;
+
+    return (value: unknown): string => {
+      const bindName = `sequelize_${++i}`;
+
+      bind[bindName] = value;
+
+      return `$${bindName}`;
+    };
+  }
+
+  /**
+   * Returns the SQL fragment to handle returning the attributes from an insert/update query.
+   *
+   * @param options         An object with options.
+   * @param modelAttributes A map with the model attributes.
+  */
+  getReturnFields(options: GetReturnFieldsOptions, modelAttributes: Map<string, AttributeOptions>): string[] {
+    const returnFields: string[] = [];
+
+    if (Array.isArray(options.returning)) {
+      returnFields.push(...options.returning.map(field => {
+        if (typeof field === 'string') {
+          return this.queryGenerator.quoteIdentifier(field);
+        } else if (field instanceof Literal) {
+          return this.queryGenerator.formatSqlExpression(field);
+        } else if (field instanceof Col) {
+          return this.queryGenerator.formatSqlExpression(field);
+        }
+
+        throw new Error(`Unsupported value in "returning" option: ${inspect(field)}. This option only accepts true, false, or an array of strings, col() or literal().`);
+      }));
+    } else if (modelAttributes.size > 0) {
+      const attributes = [...modelAttributes.entries()]
```

Comment:
this will create many arrays. To improve the perfs of Sequelize, I propose to use the functions in `iterators.ts` instead of array methods

Similarly, I recommend using `.keys()` instead of `.entries()` as from previous tests, accessing the key then the value using the key has a lower perf impact than creating an array for each entry

Once the native Iterator helpers are implemented in node, we'll be able to move to that to have a nicer code & still keep the performance improvements

Reply:
Ok, I've changed to the iterator's and also added a filter function, LMK if it looks OK.